[#11521] docs(glue): Add AWS Glue catalog and connector documentation (#11523)
apache/gravitino: docs/spark-connector/spark-catalog-glue.md

---
title: "Spark connector Glue catalog"
slug: /spark-connector/spark-catalog-glue
keyword: spark connector glue catalog aws
license: "This software is licensed under the Apache License version 2."
---

> With the Apache Gravitino Spark connector, accessing data or managing metadata in AWS Glue Data Catalog becomes straightforward, enabling seamless federation queries across Glue catalogs.

Reviewer: **[style]** This body prose appears before the first H2 heading, creating orphan content (STYLE.md: "Every doc page opens with an H2 heading after the frontmatter"). Move this sentence inside `## Introduction` or another opening H2. Also: `federation queries` → `federated queries`.

> With the Apache Gravitino Spark connector, accessing data or managing metadata in AWS Glue Data Catalog becomes straightforward, enabling seamless federation queries across Glue catalogs.

Reviewer: **[style]** This body prose appears before the first H2 heading, creating orphan content (STYLE.md: "Every doc page opens with an H2 heading after the frontmatter"). Move this sentence inside `## Introduction` or another opening H2. Also: `federation queries` → `federated queries`.

> With the Apache Gravitino Spark connector, accessing data or managing metadata in AWS Glue Data Catalog becomes straightforward, enabling seamless federation queries across Glue catalogs.

Reviewer: **[style]** This body prose appears before the first H2 heading, creating orphan content (STYLE.md: "Every doc page opens with an H2 heading after the frontmatter"). Move this sentence inside `## Introduction` or another opening H2. Also: `federation queries` → `federated queries`.

> With the Apache Gravitino Spark connector, accessing data or managing metadata in AWS Glue Data Catalog becomes straightforward, enabling seamless federation queries across Glue catalogs.

Reviewer: **[style]** This body prose appears before the first H2 heading, creating orphan content. Move this sentence inside `## Introduction` or another opening H2. Also: `federation queries` -> `federated queries`.

## Capabilities

> Supports most DDL and DML operations in SparkSQL, except such operations:

Reviewer: **[terminology]** The product name is `Spark SQL` (two words), per Apache Spark documentation. Also `except such operations` is awkward; prefer `except the following operations`.

```suggestion
Supports most DDL and DML operations in Spark SQL, except the following operations:
```

> Supports most DDL and DML operations in SparkSQL, except such operations:

Reviewer: **[terminology]** The product name is `Spark SQL` (two words), per Apache Spark documentation. Also `except such operations` is awkward; prefer `except the following operations`.

```suggestion
Supports most DDL and DML operations in Spark SQL, except the following operations:
```

> Supports most DDL and DML operations in SparkSQL, except such operations:

Reviewer: **[terminology]** The product name is `Spark SQL` (two words), per Apache Spark documentation. Also `except such operations` is awkward; prefer `except the following operations`.

```suggestion
Supports most DDL and DML operations in Spark SQL, except the following operations:
```

> Supports most DDL and DML operations in SparkSQL, except such operations:

Reviewer: **[terminology]** The product name is `Spark SQL` (two words). Also `except such operations` is awkward; prefer `except the following operations`.

```suggestion
Supports most DDL and DML operations in Spark SQL, except the following operations:
```

- Function operations (Gravitino UDFs are supported, see [Spark connector - User-defined functions](spark-connector-udf.md))
- Partition operations
- View operations
- `LOAD` clause
- `CREATE TABLE LIKE` clause
- `TRUNCATE TABLE` clause

> ## Table format support

Reviewer: **[style]** Drop the filler `support` suffix; use the noun-phrase heading. Fix Title Case.

```suggestion
## Table Formats
```

> ## Table format support

Reviewer: **[style]** Drop the filler `support` suffix; use the noun-phrase heading. Fix Title Case.

```suggestion
## Table Formats
```

> ## Table format support

Reviewer: **[style]** Drop the filler `support` suffix; use the noun-phrase heading. Fix Title Case.

```suggestion
## Table Formats
```

> ## Table format support

Reviewer: **[style]** Drop the filler `support` suffix; use the noun-phrase heading. Fix Title Case.

```suggestion
## Table Formats
```

The Glue catalog supports mixed table formats within a single database:

- **Hive-format tables** (PARQUET, ORC, TEXTFILE, etc.): routed to HiveTableCatalog using the AWS Glue Data Catalog Hive client
- **Iceberg-format tables**: routed to Iceberg's SparkCatalog (GlueCatalog) for I/O

Table routing is based on the `table-format` property in Glue table parameters. Tables with `table-format=ICEBERG` are delegated to the Iceberg backend.

## Requirements

- Network access to the AWS Glue API and Amazon S3
- AWS IAM credentials with necessary Glue and S3 permissions.
  See [AWS IAM permissions](../aws-glue-catalog.md#aws-iam-permissions) for the required policy
- Apache Spark 3.3, 3.4, or 3.5
- Patched Hive and AWS Glue client JARs (see [Setup](#setup))

## Setup

Spark's bundled Hive 2.3.9 does not include the `HiveMetaStoreClientFactory` interface
(added by [HIVE-12679](https://issues.apache.org/jira/browse/HIVE-12679)) that the AWS Glue client
requires. You must replace the bundled Hive JARs with patched versions bundled together with the

> Glue client.

Reviewer: **[style]** Drop `You must` scaffolding; use the imperative.

```suggestion
Replace the bundled Hive JARs with patched versions that include the Glue client.
```

> Glue client.

Reviewer: **[style]** Drop `You must` scaffolding; use the imperative.

```suggestion
Replace the bundled Hive JARs with patched versions that include the Glue client.
```

> Glue client.

Reviewer: **[style]** Drop `You must` scaffolding; use the imperative.

```suggestion
Replace the bundled Hive JARs with patched versions that include the Glue client.
```

> Glue client.

Reviewer: **[style]** Drop `You must` scaffolding; use the imperative.

```suggestion
Replace the bundled Hive JARs with patched versions that include the Glue client.
```

:::note
On AWS managed Spark environments such as Amazon EMR, the Hive libraries are already patched and

> the AWS Glue Data Catalog client is pre-installed. You can skip Steps 1 and 2 below and proceed

Reviewer: **[grammar]** `AWS managed` is a compound modifier before a noun and requires a hyphen.

```suggestion
On AWS-managed Spark environments such as Amazon EMR, the Hive libraries are already patched
```

> the AWS Glue Data Catalog client is pre-installed. You can skip Steps 1 and 2 below and proceed

Reviewer: **[grammar]** `AWS managed` is a compound modifier before a noun and requires a hyphen.

```suggestion
On AWS-managed Spark environments such as Amazon EMR, the Hive libraries are already patched
```

> the AWS Glue Data Catalog client is pre-installed. You can skip Steps 1 and 2 below and proceed

Reviewer: **[grammar]** `AWS managed` is a compound modifier before a noun and requires a hyphen.

```suggestion
On AWS-managed Spark environments such as Amazon EMR, the Hive libraries are already patched
```

> the AWS Glue Data Catalog client is pre-installed. You can skip Steps 1 and 2 below and proceed

Reviewer: **[grammar]** `AWS managed` is a compound modifier before a noun and requires a hyphen.

```suggestion
On AWS-managed Spark environments such as Amazon EMR, the Hive libraries are already patched
```

```bash
BASE=https://raw.githubusercontent.com/datastrato/spark-hive-glue-libs/main/spark3/glue-3.4.0
mkdir -p /opt/glue-hive-jars
for jar in \
  aws-glue-datacatalog-spark-client-3.4.0.jar \
  hive-exec-2.3.10.jar \
  hive-metastore-2.3.10.jar \
  hive-common-2.3.10.jar \
  hive-serde-2.3.10.jar \
  hive-shims-2.3.10.jar \
  aws-java-sdk-glue-1.12.31.jar \
  aws-java-sdk-core-1.12.31.jar \
  jmespath-java-1.12.31.jar; do
  wget "$BASE/$jar" -P /opt/glue-hive-jars/
done
```

### Step 2: Configure Spark

Add the following configurations when starting Spark:

```bash
spark-submit \
  --conf spark.sql.hive.metastore.version=2.3.10 \
  --conf spark.sql.hive.metastore.jars=path \
  --conf spark.sql.hive.metastore.jars.path=/opt/glue-hive-jars/* \
  --conf spark.sql.hive.metastore.sharedPrefixes=com.amazonaws,org.apache.thrift,org.slf4j,com.google.common \
  ...
```

:::note
The `spark.sql.hive.metastore.jars=path` configuration instructs Spark to load the Hive metastore
client from the specified directory instead of its bundled Hive JARs.
The AWS SDK JARs in the directory are loaded in Spark's `IsolatedClientLoader` to prevent version
conflicts with `hadoop-aws`.
:::

> ## Creating a catalog

Reviewer: **[style]** Gerund-led headings are forbidden. Use the imperative, and fix Title Case.

```suggestion
## Create a Catalog
```

> ## Creating a catalog

Reviewer: **[style]** Gerund-led headings are forbidden. Use the imperative, and fix Title Case.

```suggestion
## Create a Catalog
```

> ## Creating a catalog

Reviewer: **[style]** Gerund-led headings are forbidden. Use the imperative, and fix Title Case.

```suggestion
## Create a Catalog
```

> ## Creating a catalog

Reviewer: **[style]** Gerund-led headings are forbidden. Use the imperative, and fix Title Case.

```suggestion
## Create a Catalog
```

Use the Gravitino REST API or the Gravitino CLI to create a Glue catalog:

```bash
curl -X POST -H "Content-Type: application/json" \
-d '{
  "name": "glue_catalog",
  "type": "RELATIONAL",
  "provider": "glue",
  "properties": {
    "aws-region": "us-east-1",
    "aws-access-key-id": "<your-access-key-id>",
    "aws-secret-access-key": "<your-secret-access-key>",
    "warehouse": "s3://my-bucket/warehouse"
  }
}' http://gravitino-host:8090/api/metalakes/{metalake}/catalogs
```

For more information about the Glue catalog properties, refer to [AWS Glue catalog](../aws-glue-catalog.md).

> ## SQL example

Reviewer: **[style]** Title Case: `example` → `Example`.

```suggestion
## SQL Example
```

> ## SQL example

Reviewer: **[style]** Title Case: `example` → `Example`.

```suggestion
## SQL Example
```

> ## SQL example

Reviewer: **[style]** Title Case: `example` → `Example`.

```suggestion
## SQL Example
```

> ## SQL example

Reviewer: **[style]** Title Case: `example` -> `Example`.

```suggestion
## SQL Example
```

```sql
-- Suppose glue_catalog is the Glue catalog name managed by Gravitino
USE glue_catalog;

CREATE DATABASE IF NOT EXISTS mydatabase;
USE mydatabase;

-- Create a Hive-format Parquet table
CREATE TABLE IF NOT EXISTS employees (
    id INT,
    name STRING,
    age INT
)
PARTITIONED BY (department STRING)
STORED AS PARQUET;

DESC TABLE EXTENDED employees;

INSERT OVERWRITE TABLE employees PARTITION(department='Engineering')
VALUES (1, 'John Doe', 30), (2, 'Jane Smith', 28);

SELECT * FROM employees WHERE department = 'Engineering';

-- Create an Iceberg table partitioned by order_ts (identity)
CREATE TABLE IF NOT EXISTS orders (
    order_id BIGINT,
    customer STRING,
    amount DECIMAL(10, 2),
    order_ts TIMESTAMP
)
USING iceberg
PARTITIONED BY (order_ts);

INSERT INTO orders VALUES (1, 'alice', 99.99, TIMESTAMP '2024-01-01 00:00:00');

SELECT * FROM orders WHERE order_ts >= DATE '2024-01-01';
```

:::note

> Time-based partition transforms (`days()`, `hours()`, `months()`, `years()`) can be specified in `CREATE TABLE`, but `INSERT` into such Iceberg tables is not currently supported through the Gravitino Spark Glue connector. Use identity partitioning (e.g. `PARTITIONED BY (order_ts)`) when writes are required.

Reviewer: **[style]** Remove `currently` — docs describe current state by default (STYLE.md: "Avoid 'Currently' and 'Current' as filler").

```suggestion
but `INSERT` into such Iceberg tables is not supported through the Gravitino Spark Glue connector.
```

> Time-based partition transforms (`days()`, `hours()`, `months()`, `years()`) can be specified in `CREATE TABLE`, but `INSERT` into such Iceberg tables is not currently supported through the Gravitino Spark Glue connector. Use identity partitioning (e.g. `PARTITIONED BY (order_ts)`) when writes are required.

Reviewer: **[style]** Remove `currently` — docs describe current state by default (STYLE.md: "Avoid 'Currently' and 'Current' as filler").

```suggestion
but `INSERT` into such Iceberg tables is not supported through the Gravitino Spark Glue connector.
```

> Time-based partition transforms (`days()`, `hours()`, `months()`, `years()`) can be specified in `CREATE TABLE`, but `INSERT` into such Iceberg tables is not currently supported through the Gravitino Spark Glue connector. Use identity partitioning (e.g. `PARTITIONED BY (order_ts)`) when writes are required.

Reviewer: **[style]** Remove `currently` — docs describe current state by default (STYLE.md: "Avoid 'Currently' and 'Current' as filler").

```suggestion
but `INSERT` into such Iceberg tables is not supported through the Gravitino Spark Glue connector.
```

> Time-based partition transforms (`days()`, `hours()`, `months()`, `years()`) can be specified in `CREATE TABLE`, but `INSERT` into such Iceberg tables is not currently supported through the Gravitino Spark Glue connector. Use identity partitioning (e.g. `PARTITIONED BY (order_ts)`) when writes are required.

Reviewer: **[style]** Remove `currently` -- docs describe current state by default.

```suggestion
but `INSERT` into such Iceberg tables is not supported through the Gravitino Spark Glue connector.
```

:::

> ## Catalog properties

Reviewer: **[style]** Title Case: `properties` → `Properties`.

```suggestion
## Catalog Properties
```

> ## Catalog properties

Reviewer: **[style]** Title Case: `properties` → `Properties`.

```suggestion
## Catalog Properties
```

> ## Catalog properties

Reviewer: **[style]** Title Case: `properties` → `Properties`.

```suggestion
## Catalog Properties
```

> ## Catalog properties

Reviewer: **[style]** Title Case: `properties` -> `Properties`.

```suggestion
## Catalog Properties
```

> Gravitino Spark connector transforms the following property names defined in catalog properties to Spark Hive/Iceberg connector configuration.

Reviewer: **[clarity]** Missing `The`; `transforms...to` is awkward for property name remapping — `maps...to` is standard terminology.

```suggestion
The Gravitino Spark connector maps the following catalog property names to Spark Hive/Iceberg connector configuration.
```

> Gravitino Spark connector transforms the following property names defined in catalog properties to Spark Hive/Iceberg connector configuration.

Reviewer: **[clarity]** Missing `The`; `transforms...to` is awkward for property name remapping — `maps...to` is standard terminology.

```suggestion
The Gravitino Spark connector maps the following catalog property names to Spark Hive/Iceberg connector configuration.
```

> Gravitino Spark connector transforms the following property names defined in catalog properties to Spark Hive/Iceberg connector configuration.

Reviewer: **[clarity]** Missing `The`; `transforms...to` is awkward for property name remapping — `maps...to` is standard terminology.

```suggestion
The Gravitino Spark connector maps the following catalog property names to Spark Hive/Iceberg connector configuration.
```

> Gravitino Spark connector transforms the following property names defined in catalog properties to Spark Hive/Iceberg connector configuration.

Reviewer: **[clarity]** Missing `The`; `transforms...to` is awkward -- `maps...to` is standard terminology.

```suggestion
The Gravitino Spark connector maps the following catalog property names to Spark Hive/Iceberg connector configuration.
```

For Iceberg tables, Gravitino properties are mapped to Iceberg GlueCatalog configuration:

| Gravitino property      | Iceberg GlueCatalog property          | Description                               |
|-------------------------|---------------------------------------|-------------------------------------------|
| `aws-region`            | `client.region`                       | AWS region                                |
| `aws-glue-catalog-id`   | `glue.id`                             | Glue catalog ID                           |
| `aws-glue-endpoint`     | `glue.endpoint`                       | Glue endpoint URL                         |
| `aws-access-key-id`     | `client.credentials-provider.*`      | AWS access key (via custom provider)      |
| `aws-secret-access-key` | `client.credentials-provider.*`      | AWS secret key (via custom provider)      |

Gravitino catalog property names with the prefix `spark.bypass.` are passed directly to the Spark Hive connector. For example, using `spark.bypass.hive.exec.dynamic.partition.mode` to pass `hive.exec.dynamic.partition.mode` to the Spark Hive connector.

> ## S3 storage

Reviewer: **[style]** Title Case: `storage` → `Storage`.

```suggestion
## S3 Storage
```

> ## S3 storage

Reviewer: **[style]** Title Case: `storage` → `Storage`.

```suggestion
## S3 Storage
```

> ## S3 storage

Reviewer: **[style]** Title Case: `storage` → `Storage`.

```suggestion
## S3 Storage
```

> ## S3 storage

Reviewer: **[style]** Title Case: `storage` -> `Storage`.

```suggestion
## S3 Storage
```
